src/pykeen/models/predict.py

```diff
@@ -115,6 +114,143 @@ def _get_input_batch(
+            df[f"in_{key}"] = self._contains(df=df, mapped_triples=mapped_triples)
+        return df
+
+    def process(self, df: pd.DataFrame, remove_known: bool, add_novelties: bool) -> pd.DataFrame:
+        """Post-process a prediction dataframe."""
+        if add_novelties and not remove_known:
+            return self.contains(df=df)
+        elif remove_known:
+            return self.filter(df=df)
+        return df
+
+
+class SinglePredictionPostProcessor(PredictionPostProcessor):
+    """Post-processor for single-target predictions."""
+
+    def __init__(self, target: Target, other_columns_fixed_ids: Tuple[int, int], **kwargs) -> None:
+        """Initialize the post-processor.
+
+        :param target:
+            the prediction target
+        :param other_columns_fixed_ids:
+            the fixed IDs for the other columns
+
+        :param kwargs:
+            additional keyword-based parameters passed to :meth:`PredictionPostProcessor.__init__`
+        """
+        super().__init__(**kwargs)
+        self.target = target
+        self.other_columns_fixed_ids = other_columns_fixed_ids
+
+    # docstr-coverage: inherited
+    def _contains(
+        self, df: pd.DataFrame, mapped_triples: MappedTriples, invert: bool = False
+    ) -> numpy.ndarray:  # noqa: D102
+        col = TARGET_TO_INDEX[self.target]
+        other_cols = sorted(set(range(mapped_triples.shape[1])).difference({col}))
+        device = mapped_triples.device
+        other_col_ids = torch.as_tensor(data=self.other_columns_fixed_ids, dtype=torch.long, device=device)
+        filter_mask = (mapped_triples[:, other_cols] == other_col_ids[None, :]).all(dim=-1)
+        known_ids = mapped_triples[filter_mask, col].unique()
+        query_ids = torch.as_tensor(df[f"{self.target}_id"].to_numpy(), device=mapped_triples.device)
+        return torch.isin(elements=query_ids, test_elements=known_ids, assume_unique=True, invert=invert).cpu().numpy()
+
+
+def isin_many_dim(elements: torch.Tensor, test_elements: torch.Tensor, dim: int = 0) -> torch.BoolTensor:
+    """Return whether elements are contained in test elements."""
+    inverse, counts = torch.cat([elements, test_elements], dim=dim).unique(
+        return_counts=True, return_inverse=True, dim=dim
+    )[1:]
+    return counts[inverse[: elements.shape[dim]]] > 1
+
+
+class AllPredictionPostProcessor(PredictionPostProcessor):
+    """Post-processor for all-triples predictions."""
+
+    # docstr-coverage: inherited
+    def _contains(
+        self, df: pd.DataFrame, mapped_triples: MappedTriples, invert: bool = False
+    ) -> numpy.ndarray:  # noqa: D102
+        contained = (
+            isin_many_dim(
+                elements=torch.as_tensor(
+                    df[[f"{target}_id" for target, _ in sorted(TARGET_TO_INDEX.items(), key=itemgetter(1))]].values,
+                    device=mapped_triples.device,
+                ),
+                test_elements=mapped_triples,
+            )
+            .cpu()
```

Comment:
maybe better to return the torch tensor then just cpu().numpy() in the place where it gets used?

Reply:
Currently, we need `cpu().numpy()` everytime we call this function, so I would leave it as it is for now.